File: cocos/editor-support/cocostudio/CCActionNode.cpp
/****************************************************************************
Copyright (c) 2013-2016 Chukong Technologies Inc.
Copyright (c) 2017-2018 Xiamen Yaji Software Co., Ltd.

http://www.cocos2d-x.org

Permission is hereby granted, free of charge, to any person obtaining a copy
of this software and associated documentation files (the "Software"), to deal
in the Software without restriction, including without limitation the rights
to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
copies of the Software, and to permit persons to whom the Software is
furnished to do so, subject to the following conditions:

The above copyright notice and this permission notice shall be included in
all copies or substantial portions of the Software.

THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
THE SOFTWARE.
****************************************************************************/

#include "editor-support/cocostudio/CCActionNode.h"
#include "editor-support/cocostudio/CCActionFrameEasing.h"
#include "ui/UIWidget.h"
#include "ui/UIHelper.h"
#include "ui/UILayout.h"
#include "editor-support/cocostudio/CocoLoader.h"
#include "base/ccUtils.h"
#include "editor-support/cocostudio/CCActionManagerEx.h"


using namespace cocos2d;
using namespace ui;

namespace cocostudio {

ActionNode::ActionNode()
: _currentFrameIndex(0)
, _destFrameIndex(0)
, _fUnitTime(0.1f)
, _actionTag(0)
, _actionSpawn(nullptr)
, _action(nullptr)
, _object(nullptr)
, _frameArrayNum(0)
{
    _frameArrayNum = (int)kKeyframeMax;
    for(int i = 0; i < _frameArrayNum; i++)
    {
        _frameArray.push_back( new cocos2d::Vector<ActionFrame*>());
    }
}

ActionNode::~ActionNode()
{
    if (_action == nullptr)
    {
        CC_SAFE_RELEASE_NULL(_actionSpawn);
    }
    else
    {
        CC_SAFE_RELEASE_NULL(_action);
        CC_SAFE_RELEASE_NULL(_actionSpawn);
    }
    
    CC_SAFE_RELEASE(_object);

    for (auto object : _frameArray)
    {
        object->clear();
        CC_SAFE_DELETE(object);
    }
    _frameArray.clear();
}

void ActionNode::initWithDictionary(const rapidjson::Value& dic, Ref* root)
{
    Widget * rw = dynamic_cast<Widget *>(root);
    if (nullptr == rw)
        return;

    setActionTag(DICTOOL->getIntValue_json(dic, "ActionTag"));
    Widget* node = Helper::seekActionWidgetByActionTag(rw, getActionTag());
    bool positionOffset = node && (nullptr == (dynamic_cast<Layout *>(node)));

    int actionFrameCount = DICTOOL->getArrayCount_json(dic, "actionframelist");
    for (int i=0; i<actionFrameCount; i++)
    {
        const rapidjson::Value& actionFrameDic = DICTOOL->getDictionaryFromArray_json(dic, "actionframelist", i);
        int frameInex = DICTOOL->getIntValue_json(actionFrameDic,"frameid");

        int frameTweenType = DICTOOL->getIntValue_json(actionFrameDic,"tweenType");

        std::vector<float> frameTweenParameter;
        int frameTweenParameterNum = DICTOOL->getArrayCount_json(actionFrameDic, "tweenParameter");

        for (int j = 0; j < frameTweenParameterNum; j++)
        {
            float value = DICTOOL->getFloatValueFromArray_json(actionFrameDic, "tweenParameter", j);
            frameTweenParameter.push_back(value);
        }


        bool existPosition = DICTOOL->checkObjectExist_json(actionFrameDic,"positionx");
        if (existPosition)
        {
            float positionX = DICTOOL->getFloatValue_json(actionFrameDic, "positionx");
            float positionY = DICTOOL->getFloatValue_json(actionFrameDic, "positiony");
            if (positionOffset && (nullptr != node->getParent()) && ActionManagerEx::getInstance()->getStudioVersionNumber() < 1600)
            {
                Vec2 AnchorPointIn = node->getParent()->getAnchorPointInPoints();
                positionX += AnchorPointIn.x;
                positionY += AnchorPointIn.y;
            }
            ActionMoveFrame* actionFrame = new (std::nothrow) ActionMoveFrame();
            actionFrame->setFrameIndex(frameInex);
            actionFrame->setEasingType(frameTweenType);
            actionFrame->setEasingParameter(frameTweenParameter);
            actionFrame->setPosition(Vec2(positionX, positionY));
            auto cActionArray = _frameArray.at((int)kKeyframeMove);
            cActionArray->pushBack(actionFrame);
            actionFrame->release();
        }

        bool existScale = DICTOOL->checkObjectExist_json(actionFrameDic,"scalex");
        if (existScale)
        {
            float scaleX = DICTOOL->getFloatValue_json(actionFrameDic, "scalex");
            float scaleY = DICTOOL->getFloatValue_json(actionFrameDic, "scaley");
            ActionScaleFrame* actionFrame = new (std::nothrow) ActionScaleFrame();
            actionFrame->setFrameIndex(frameInex);
            actionFrame->setEasingType(frameTweenType);
            actionFrame->setEasingParameter(frameTweenParameter);
            actionFrame->setScaleX(scaleX);
            actionFrame->setScaleY(scaleY);
            auto cActionArray = _frameArray.at((int)kKeyframeScale);
            cActionArray->pushBack(actionFrame);
            actionFrame->release();             
        }

        bool existRotation = DICTOOL->checkObjectExist_json(actionFrameDic,"rotation");
        if (existRotation)
        {
            float rotation = DICTOOL->getFloatValue_json(actionFrameDic, "rotation");
            ActionRotationFrame* actionFrame = new (std::nothrow) ActionRotationFrame();
            actionFrame->setFrameIndex(frameInex);
            actionFrame->setEasingType(frameTweenType);
            actionFrame->setEasingParameter(frameTweenParameter);
            actionFrame->setRotation(rotation);
            auto cActionArray = _frameArray.at((int)kKeyframeRotate);
            cActionArray->pushBack(actionFrame);
            actionFrame->release();
        }

        bool existOpacity = DICTOOL->checkObjectExist_json(actionFrameDic,"opacity");
        if (existOpacity)
        {
            int opacity = DICTOOL->getIntValue_json(actionFrameDic, "opacity");
            ActionFadeFrame* actionFrame = new (std::nothrow) ActionFadeFrame();
            actionFrame->setFrameIndex(frameInex);
            actionFrame->setEasingType(frameTweenType);
            actionFrame->setEasingParameter(frameTweenParameter);
            actionFrame->setOpacity(opacity);
            auto cActionArray = _frameArray.at((int)kKeyframeTint);
            cActionArray->pushBack(actionFrame);
            actionFrame->release();
        }

        bool existColor = DICTOOL->checkObjectExist_json(actionFrameDic,"colorr");
        if (existColor)
        {
            int colorR = DICTOOL->getIntValue_json(actionFrameDic, "colorr");
            int colorG = DICTOOL->getIntValue_json(actionFrameDic, "colorg");
            int colorB = DICTOOL->getIntValue_json(actionFrameDic, "colorb");
            ActionTintFrame* actionFrame = new (std::nothrow) ActionTintFrame();
            actionFrame->setFrameIndex(frameInex);
            actionFrame->setEasingType(frameTweenType);
            actionFrame->setEasingParameter(frameTweenParameter);
            actionFrame->setColor(Color3B(colorR,colorG,colorB));
            auto cActionArray = _frameArray.at((int)kKeyframeFade);
            cActionArray->pushBack(actionFrame);
            actionFrame->release();
        }
    }
    initActionNodeFromRoot(root);
}
    
    int ActionNode::valueToInt(const std::string& value)
    {
        return atoi(value.c_str());
    }
    bool ActionNode::valueToBool(const std::string& value)
    {
        int intValue = valueToInt(value);
        if (1 == intValue) {
            return true;
        }else{
            return false;
        }
    }
    float ActionNode::valueToFloat(const std::string& value)
    {
        return utils::atof(value.c_str());
    }
    
    void ActionNode::initWithBinary(CocoLoader *cocoLoader,
                                    stExpCocoNode *cocoNode,
                                    cocos2d::Ref *root)
    {
        
        stExpCocoNode *stChildNode = cocoNode;
        
        int actionNodeCount =  stChildNode->GetChildNum();
        stChildNode = stChildNode[0].GetChildArray(cocoLoader);
        stExpCocoNode *frameListNode = nullptr;
        for (int i = 0; i < actionNodeCount; ++i) {
            std::string key = stChildNode[i].GetName(cocoLoader);
            std::string value = stChildNode[i].GetValue(cocoLoader);
            if (key == "ActionTag") {
                setActionTag(valueToInt(value));
            }else if (key == "actionframelist"){
                frameListNode = &stChildNode[i];
            }
        }
        
        int actionFrameCount = frameListNode->GetChildNum();
        stExpCocoNode *stFrameChildNode = frameListNode->GetChildArray(cocoLoader);
        for (int i=0; i<actionFrameCount; i++) {
            
            int frameIndex;
            int frameTweenType;
            float positionX;
            float positionY;
            float scaleX;
            float scaleY;
            float rotation;
            int opacity;
            int colorR = -1;
            int colorG = -1;
            int colorB = -1;
            std::vector<float> frameTweenParameter;
            
            int framesCount = stFrameChildNode[i].GetChildNum();
            stExpCocoNode *innerFrameNode = stFrameChildNode[i].GetChildArray(cocoLoader);
            for (int j = 0; j < framesCount; j++) {
                std::string key = innerFrameNode[j].GetName(cocoLoader);
                std::string value = innerFrameNode[j].GetValue(cocoLoader);
                
                if (key == "frameid") {
                    frameIndex = valueToInt(value);
                }else if(key == "tweenType"){
                    frameTweenType = valueToInt(value);
                }else if (key == "tweenParameter"){
                    //  There are no tweenParameter args in the json file
                    int tweenParameterCount = innerFrameNode[j].GetChildNum();
                    stExpCocoNode *tweenParameterArray = innerFrameNode[j].GetChildArray(cocoLoader);
                    for (int k = 0; k < tweenParameterCount; ++k) {
                        std::string t_key = tweenParameterArray[j].GetName(cocoLoader);
                        std::string t_value = tweenParameterArray[j].GetValue(cocoLoader);
                        frameTweenParameter.push_back(valueToFloat(t_value));
                    }
                }else if (key == "positionx"){
                    positionX = valueToFloat(value);
                }else if (key == "positiony"){
                    positionY = valueToFloat(value);
                    ActionMoveFrame* actionFrame = new (std::nothrow) ActionMoveFrame();
                    actionFrame->autorelease();
                    actionFrame->setEasingType(frameTweenType);
                    actionFrame->setEasingParameter(frameTweenParameter);
                    actionFrame->setFrameIndex(frameIndex);
                    actionFrame->setPosition(Vec2(positionX, positionY));
                    auto cActionArray = _frameArray.at((int)kKeyframeMove);
                    cActionArray->pushBack(actionFrame);
                }else if(key == "scalex"){
                    scaleX = valueToFloat(value);
                }else if(key == "scaley"){
                    scaleY = valueToFloat(value);
                    ActionScaleFrame* actionFrame = new (std::nothrow) ActionScaleFrame();
                    actionFrame->autorelease();
                    actionFrame->setEasingType(frameTweenType);
                    actionFrame->setEasingParameter(frameTweenParameter);
                    actionFrame->setFrameIndex(frameIndex);
                    actionFrame->setScaleX(scaleX);
                    actionFrame->setScaleY(scaleY);
                    auto cActionArray = _frameArray.at((int)kKeyframeScale);
                    cActionArray->pushBack(actionFrame);
                }else if (key == "rotation"){
                    rotation = valueToFloat(value);
                    ActionRotationFrame* actionFrame = new (std::nothrow) ActionRotationFrame();
                    actionFrame->autorelease();
                    actionFrame->setEasingType(frameTweenType);
                    actionFrame->setEasingParameter(frameTweenParameter);
                    actionFrame->setFrameIndex(frameIndex);
                    actionFrame->setRotation(rotation);
                    auto cActionArray = _frameArray.at((int)kKeyframeRotate);
                    cActionArray->pushBack(actionFrame);
                }else if (key == "opacity"){
                    opacity = valueToInt(value);
                    ActionFadeFrame* actionFrame = new (std::nothrow) ActionFadeFrame();
                    actionFrame->autorelease();
                    actionFrame->setEasingType(frameTweenType);
                    actionFrame->setEasingParameter(frameTweenParameter);
                    actionFrame->setFrameIndex(frameIndex);
                    actionFrame->setOpacity(opacity);
                    auto cActionArray = _frameArray.at((int)kKeyframeFade);
                    cActionArray->pushBack(actionFrame);
                }else if (key == "colorb"){
                    colorB = valueToInt(value);
                }else if(key == "colorg"){
                    colorG = valueToInt(value);
                }else if(key == "colorr"){
                    colorR = valueToInt(value);
                    
                    ActionTintFrame* actionFrame = new (std::nothrow) ActionTintFrame();
                    actionFrame->autorelease();
                    actionFrame->setEasingType(frameTweenType);
                    actionFrame->setEasingParameter(frameTweenParameter);
                    actionFrame->setFrameIndex(frameIndex);
                    actionFrame->setColor(Color3B(colorR,colorG,colorB));
                    auto cActionArray = _frameArray.at((int)kKeyframeTint);
                    cActionArray->pushBack(actionFrame);
                }
                
                
            }
            
        }
        initActionNodeFromRoot(root);
    }

void ActionNode::initActionNodeFromRoot(Ref* root)
{
    Widget* rootWidget = dynamic_cast<Widget*>(root);
    if (rootWidget != nullptr)
    {
        Widget* widget = Helper::seekActionWidgetByActionTag(rootWidget, getActionTag());
        if (widget != nullptr)
        {
            setObject(widget);
        }
    }
}

void ActionNode::setUnitTime(float fTime)
{
    _fUnitTime = fTime;
    this->refreshActionProperty();
}

float ActionNode::getUnitTime()
{
    return _fUnitTime;
}

void ActionNode::setActionTag(int tag)
{
    _actionTag = tag;
}

int ActionNode::getActionTag()
{
    return _actionTag;
}

void ActionNode::setObject(Ref* node)
{
    CC_SAFE_RELEASE(_object);
    _object = node;
    CC_SAFE_RETAIN(_object);
}

Ref*  ActionNode::getObject()
{
    return _object;
}

Node* ActionNode::getActionNode()
{
    Node* cNode = dynamic_cast<Node*>(_object);
    if (cNode != nullptr)
    {
        return cNode;
    }
    return nullptr;
}

void ActionNode::insertFrame(int index, ActionFrame* frame)
{
    if (frame == nullptr)
    {
        return;
    }
    int frameType = frame->getFrameType();
    if(frameType < (int)_frameArray.size())
    {
        auto cArray = _frameArray.at(frameType);
        cArray->insert(index, frame);
    }
}

void ActionNode::addFrame(ActionFrame* frame)
{
    if (frame == nullptr)
    {
        return;
    }
    int frameType = frame->getFrameType();

    if(frameType < (int)_frameArray.size())
    {
        auto cArray = _frameArray.at(frameType);
        cArray->pushBack(frame);
    }
}

void ActionNode::deleteFrame(ActionFrame* frame)
{
    if (frame == nullptr)
    {
        return;
    }
    int frameType = frame->getFrameType();
    if(frameType < (int)_frameArray.size())
    {
        auto cArray = _frameArray.at(frameType);
        cArray->eraseObject(frame);
    }
}

void ActionNode::clearAllFrame()
{
    for(auto array : _frameArray)
    {
        array->clear();
    }
}

Spawn * ActionNode::refreshActionProperty()
{
    if ( _object == nullptr )
    {
        return nullptr;
    }
    Vector<FiniteTimeAction*> cSpawnArray;

    for (int n = 0; n < _frameArrayNum; n++)
    {
        auto cArray = _frameArray.at(n);
        if (cArray->size() <= 0)
        {
            continue;
        }

        Vector<FiniteTimeAction*> cSequenceArray;
        auto frameCount = cArray->size();
		if(frameCount > 1)
 		{ 
 			for (int i = 0; i < frameCount; i++)
 			{
 				auto frame = cArray->at(i);
 				if (i == 0)
 				{
// #11173 Fixed every node of UI animation(json) is starting at frame 0.                     
//                  if (frame->getFrameIndex() > 0)
//				    {
//					    DelayTime* cDelayTime = DelayTime::create(frame->getFrameIndex() * getUnitTime());
//					    if (cDelayTime != nullptr)
//						    cSequenceArray.pushBack(static_cast<FiniteTimeAction*>(cDelayTime));
//				    }
 				}
 				else
 				{
 					auto srcFrame = cArray->at(i-1);
 					float duration = (frame->getFrameIndex() - srcFrame->getFrameIndex()) * getUnitTime();
 					Action* cAction = frame->getAction(duration);
 					if(cAction != nullptr)
 					cSequenceArray.pushBack(static_cast<FiniteTimeAction*>(cAction));
 				}
 			}
 		}
 		else if (frameCount == 1)
 		{
 			auto frame = cArray->at(0);
 			float duration = 0.0f;
 			Action* cAction = frame->getAction(duration);
 			if (cAction != nullptr)
 				cSequenceArray.pushBack(static_cast<FiniteTimeAction*>(cAction));
 		}
 		Sequence* cSequence = Sequence::create(cSequenceArray);
 		if (cSequence != nullptr)
 		{
 			cSpawnArray.pushBack(cSequence);
 		}
    }

    if (_action == nullptr)
    {
    CC_SAFE_RELEASE_NULL(_actionSpawn);
    }
    else
    {
    CC_SAFE_RELEASE_NULL(_action);
    CC_SAFE_RELEASE_NULL(_actionSpawn);
    }

    _actionSpawn = Spawn::create(cSpawnArray);
    CC_SAFE_RETAIN(_actionSpawn);
    return _actionSpawn;
}

void ActionNode::playAction()
{
    if ( _object == nullptr || _actionSpawn == nullptr)
    {
        return;
    }

    if (_action!=nullptr)
    {
        _action->release();
    }

    _action = Sequence::create(_actionSpawn, nullptr);
    _action->retain();

    this->runAction();

}

void ActionNode::runAction()
{
    Node* cNode = this->getActionNode();
    if (cNode != nullptr && _action != nullptr)
    {
        cNode->runAction(_action);
    }
}

void ActionNode::stopAction()
{
    Node* cNode = this->getActionNode();
    if (cNode != nullptr && _action != nullptr)
    {
        cNode->stopAction(_action);
    }
}

int ActionNode::getFirstFrameIndex()
{
    int frameindex = 99999;
    bool bFindFrame = false;
    for (int n = 0; n < _frameArrayNum; n++)
    {
        auto cArray = _frameArray.at(n);
        if (cArray->empty())
        {
            continue;
        }
        bFindFrame = true;
        auto frame = cArray->at(0);
        int iFrameIndex = frame->getFrameIndex();

        if (frameindex > iFrameIndex)
        {
            frameindex = iFrameIndex;
        }
    }
    if (!bFindFrame)
    {
        frameindex = 0;
    }
    return frameindex;
}

int ActionNode::getLastFrameIndex()
{
    int frameindex = -1;
    bool bFindFrame = false;
    for (int n = 0; n < _frameArrayNum; n++)
    {
        auto cArray = _frameArray.at(n);
        if (cArray->empty())
        {
            continue;
        }
        bFindFrame = true;
        ssize_t lastInex = cArray->size() - 1;
        auto frame = cArray->at(lastInex);
        int iFrameIndex = frame->getFrameIndex();

        if (frameindex < iFrameIndex)
        {
            frameindex = iFrameIndex;
        }
    }
    if (!bFindFrame)
    {
        frameindex = 0;
    }
    return frameindex;
}
    
bool ActionNode::updateActionToTimeLine(float fTime)
{
    bool bFindFrame = false;

    ActionFrame* srcFrame = nullptr;
    for (int n = 0; n < _frameArrayNum; n++)
    {
        auto cArray = _frameArray.at(n);
        if (cArray->empty())
        {
            continue;
        }
        ssize_t frameCount = cArray->size();
        for (int i = 0; i < frameCount; i++)
        {
            auto frame = cArray->at(i);

            if (frame->getFrameIndex()*getUnitTime() == fTime)
            {
                this->easingToFrame(1.0f,1.0f,nullptr,frame);
                bFindFrame = true;
                break;
            }
            else if (frame->getFrameIndex()*getUnitTime() > fTime)
            {
                if (i == 0)
                {
                    this->easingToFrame(1.0f,1.0f,nullptr,frame);
                    bFindFrame = false;
                }
                else
                {
                    srcFrame = cArray->at(i-1);
                    float duration = (frame->getFrameIndex() - srcFrame->getFrameIndex())*getUnitTime();
                    float delaytime = fTime - srcFrame->getFrameIndex()*getUnitTime();
                    this->easingToFrame(duration,1.0f,nullptr,srcFrame);
                    //float easingTime = ActionFrameEasing::bounceTime(delaytime);
                    this->easingToFrame(duration,delaytime/duration,srcFrame,frame);
                    bFindFrame = true;
                }
                break;
            }
        }
    }
    return bFindFrame;
}

void ActionNode::easingToFrame(float duration,float delayTime,ActionFrame* srcFrame,ActionFrame* destFrame)
{
    Action* cAction = destFrame->getAction(duration,srcFrame);
    Node* cNode = this->getActionNode();
    if (cAction == nullptr || cNode == nullptr)
    {
        return;
    }   
    cAction->startWithTarget(cNode);
    cAction->update(delayTime);
}


bool ActionNode::isActionDoneOnce()
{
    if (_action == nullptr)
    {
        return true;
    }
    return _action->isDone();
}

}

File: tools/framework-compile/bin-templates/js-template-runtime/frameworks/cocos2d-html5/extensions/cocostudio/action/CCActionNode.js
/****************************************************************************
 Copyright (c) 2011-2012 cocos2d-x.org
 Copyright (c) 2013-2014 Chukong Technologies Inc.

 http://www.cocos2d-x.org

 Permission is hereby granted, free of charge, to any person obtaining a copy
 of this software and associated documentation files (the "Software"), to deal
 in the Software without restriction, including without limitation the rights
 to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 copies of the Software, and to permit persons to whom the Software is
 furnished to do so, subject to the following conditions:

 The above copyright notice and this permission notice shall be included in
 all copies or substantial portions of the Software.

 THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 THE SOFTWARE.
 ****************************************************************************/

/**
 * The Cocostudio's action node, it contains action target, action frame list and current frame index.  it can be play action by calling playAciton.
 * @class
 * @extends ccs.Class
 */
ccs.ActionNode = ccs.Class.extend(/** @lends ccs.ActionNode# */{
    _currentFrameIndex: 0,
    _destFrameIndex: 0,
    _unitTime: 0,
    _actionTag: 0,
    _object: null,
    _actionSpawn: null,
    _action: null,
    _frameArray: null,
    _frameArrayNum: 0,

    /**
     * Construction of ccs.ActionNode
     */
    ctor: function () {
        this._currentFrameIndex = 0;
        this._destFrameIndex = 0;
        this._unitTime = 0.1;
        this._actionTag = 0;
        this._object = null;
        this._actionSpawn = null;
        this._action = null;
        this._frameArray = [];
        this._frameArrayNum = ccs.FRAME_TYPE_MAX;
        for (var i = 0; i < this._frameArrayNum; i++)
            this._frameArray.push([]);
    },

    /**
     *  Init properties with a json dictionary
     * @param {Object} dic
     * @param {Object} root
     */
    initWithDictionary: function (dic, root) {
        this.setActionTag(dic["ActionTag"]);
        var actionFrameList = dic["actionframelist"];
        for (var i = 0; i < actionFrameList.length; i++) {
            var actionFrameDic = actionFrameList[i];
            var frameIndex = actionFrameDic["frameid"];
            var frameTweenType = actionFrameDic["tweenType"];
            if(frameTweenType == null)
                frameTweenType = 0;
            var frameTweenParameterNum = actionFrameDic["tweenParameter"];

            var frameTweenParameter = [];
            for (var j = 0; j < frameTweenParameterNum; j++){
                var value = actionFrameDic["tweenParameter"][j];
                frameTweenParameter.push(value);
            }

            var actionFrame, actionArray;
            if (actionFrameDic["positionx"] !== undefined) {
                var positionX = actionFrameDic["positionx"];
                var positionY = actionFrameDic["positiony"];
                actionFrame = new ccs.ActionMoveFrame();
                actionFrame.frameIndex = frameIndex;
                actionFrame.setEasingType(frameTweenType);
                actionFrame.setEasingParameter(frameTweenParameter);
                actionFrame.setPosition(positionX, positionY);
                actionArray = this._frameArray[ccs.FRAME_TYPE_MOVE];
                actionArray.push(actionFrame);
            }

            if (actionFrameDic["scalex"] !== undefined) {
                var scaleX = actionFrameDic["scalex"];
                var scaleY = actionFrameDic["scaley"];
                actionFrame = new ccs.ActionScaleFrame();
                actionFrame.frameIndex = frameIndex;
                actionFrame.setEasingType(frameTweenType);
                actionFrame.setEasingParameter(frameTweenParameter);
                actionFrame.setScaleX(scaleX);
                actionFrame.setScaleY(scaleY);
                actionArray = this._frameArray[ccs.FRAME_TYPE_SCALE];
                actionArray.push(actionFrame);
            }

            if (actionFrameDic["rotation"] !== undefined) {
                var rotation = actionFrameDic["rotation"];
                actionFrame = new ccs.ActionRotationFrame();
                actionFrame.frameIndex = frameIndex;
                actionFrame.setEasingType(frameTweenType);
                actionFrame.setEasingParameter(frameTweenParameter);
                actionFrame.setRotation(rotation);
                actionArray = this._frameArray[ccs.FRAME_TYPE_ROTATE];
                actionArray.push(actionFrame);
            }

            if (actionFrameDic["opacity"] !== undefined) {
                var opacity = actionFrameDic["opacity"];
                actionFrame = new ccs.ActionFadeFrame();
                actionFrame.frameIndex = frameIndex;
                actionFrame.setEasingType(frameTweenType);
                actionFrame.setEasingParameter(frameTweenParameter);
                actionFrame.setOpacity(opacity);
                actionArray = this._frameArray[ccs.FRAME_TYPE_FADE];
                actionArray.push(actionFrame);
            }

            if (actionFrameDic["colorr"] !== undefined) {
                var colorR = actionFrameDic["colorr"];
                var colorG = actionFrameDic["colorg"];
                var colorB = actionFrameDic["colorb"];
                actionFrame = new ccs.ActionTintFrame();
                actionFrame.frameIndex = frameIndex;
                actionFrame.setEasingType(frameTweenType);
                actionFrame.setEasingParameter(frameTweenParameter);
                actionFrame.setColor(cc.color(colorR, colorG, colorB));
                actionArray = this._frameArray[ccs.FRAME_TYPE_TINT];
                actionArray.push(actionFrame);
            }
            actionFrameDic = null;
        }
        this._initActionNodeFromRoot(root);
    },

    _initActionNodeFromRoot: function (root) {
        if (root instanceof ccui.Widget) {
            var widget = ccui.helper.seekActionWidgetByActionTag(root, this.getActionTag());
            if (widget)
                this.setObject(widget);
        }
    },

    /**
     * Sets the time interval of frame.
     * @param {number} time
     */
    setUnitTime: function (time) {
        this._unitTime = time;
        this._refreshActionProperty();
    },

    /**
     * Returns the time interval of frame.
     * @returns {number}
     */
    getUnitTime: function () {
        return this._unitTime;
    },

    /**
     * Sets tag to ccs.ActionNode
     * @param {Number} tag
     */
    setActionTag: function (tag) {
        this._actionTag = tag;
    },

    /**
     * Returns the tag of ccs.ActionNode
     * @returns {number}
     */
    getActionTag: function () {
        return this._actionTag;
    },

    /**
     * Sets node which will run a action.
     * @param {Object} node
     */
    setObject: function (node) {
        this._object = node;
    },

    /**
     * Returns node which will run a action.
     * @returns {*}
     */
    getObject: function () {
        return this._object;
    },

    /**
     * Returns the target node of ccs.ActionNode
     * @returns {cc.Node}
     */
    getActionNode: function () {
        if (this._object instanceof cc.Node)
            return this._object;
        return null;
    },

    /**
     * Inserts an ActionFrame to ccs.ActionNode.
     * @param {number} index
     * @param {ccs.ActionFrame} frame
     */
    insertFrame: function (index, frame) {
        if (frame == null)
            return;
        var frameType = frame.frameType;
        var array = this._frameArray[frameType];
        array.splice(index, 0, frame);
    },

    /**
     * Pushes back an ActionFrame to ccs.ActionNode.
     * @param {ccs.ActionFrame} frame
     */
    addFrame: function (frame) {
        if (!frame)
            return;
        var frameType = frame.frameType;
        var array = this._frameArray[frameType];
        array.push(frame);
    },

    /**
     * Removes an ActionFrame from ccs.ActionNode.
     * @param {ccs.ActionFrame} frame
     */
    deleteFrame: function (frame) {
        if (frame == null)
            return;
        var frameType = frame.frameType;
        var array = this._frameArray[frameType];
        cc.arrayRemoveObject(array, frame);
    },

    /**
     * Removes all ActionFrames from ccs.ActionNode.
     */
    clearAllFrame: function () {
        for (var i = 0; i < this._frameArrayNum; i++)
            this._frameArray[i].length = 0;
    },

    _refreshActionProperty: function () {
        if (this._object === null)
            return null;
        var locSpawnArray = [];
        for (var i = 0; i < this._frameArrayNum; i++) {
            var locArray = this._frameArray[i];
            if (locArray.length <= 0)
                continue;
            var locSequenceArray = [];
            for (var j = 0; j < locArray.length; j++) {
                var locFrame = locArray[j];
                if (j !== 0) {
                    var locSrcFrame = locArray[j - 1];
                    var locDuration = (locFrame.frameIndex - locSrcFrame.frameIndex) * this.getUnitTime();
                    var locAction = locFrame.getAction(locDuration);
                    if(locAction)
                        locSequenceArray.push(locAction);
                }
            }
            if(locSequenceArray){
                var locSequence = cc.sequence(locSequenceArray);
                if (locSequence !== null)
                    locSpawnArray.push(locSequence);
            }
        }

        this._action = null;
        this._actionSpawn = cc.spawn(locSpawnArray);
        return this._actionSpawn;
    },

    /**
     * Plays ccs.ActionNode's action.
     * @param {cc.CallFunc} fun
     */
    playAction: function (fun) {
        if (this._object === null || this._actionSpawn === null)
            return;
        if(fun)
            this._action = cc.sequence(this._actionSpawn, fun);
        else
            this._action = cc.sequence(this._actionSpawn);
        this._runAction();
    },

    _runAction: function () {
        var node = this.getActionNode();
        if (node !== null && this._action !== null)
            node.runAction(this._action);
    },

    /**
     * Stops action.
     */
    stopAction: function () {
        var node = this.getActionNode();
        if (node !== null && this._action !== null) {
            if(!this._action.isDone())
                node.stopAction(this._action);
        }
    },

    /**
     * Returns index of first ActionFrame.
     * @returns {number}
     */
    getFirstFrameIndex: function () {
        var locFrameindex = 99999;
        var bFindFrame = false, locFrameArray = this._frameArray;
        for (var i = 0, len = this._frameArrayNum; i < len; i++) {
            var locArray = locFrameArray[i];
            if (locArray.length <= 0)
                continue;
            bFindFrame = true;
            var locFrameIndex = locArray[0].frameIndex;
            locFrameindex = locFrameindex > locFrameIndex ? locFrameIndex : locFrameindex;
        }
        if (!bFindFrame)
            locFrameindex = 0;
        return locFrameindex;
    },

    /**
     * Returns the index of last ccs.ActionFrame.
     * @returns {number}
     */
    getLastFrameIndex: function () {
        var locFrameindex = -1;
        var locIsFindFrame = false ,locFrameArray = this._frameArray;
        for (var i = 0, len = this._frameArrayNum; i < len; i++) {
            var locArray = locFrameArray[i];
            if (locArray.length <= 0)
                continue;
            locIsFindFrame = true;
            var locFrame = locArray[locArray.length - 1];
            var locFrameIndex = locFrame.frameIndex;
            locFrameindex = locFrameindex < locFrameIndex ? locFrameIndex : locFrameindex;
        }
        if (!locIsFindFrame)
            locFrameindex = 0;
        return locFrameindex;
    },

    /**
     * Updates action states to some time.
     * @param {Number} time
     * @returns {boolean}
     */
    updateActionToTimeLine: function (time) {
        var locIsFindFrame = false;
        var locUnitTime = this.getUnitTime();
        for (var i = 0; i < this._frameArrayNum; i++) {
            var locArray = this._frameArray[i];
            if (locArray === null)
                continue;

            for (var j = 0; j < locArray.length; j++) {
                var locFrame = locArray[j];
                if (locFrame.frameIndex * locUnitTime === time) {
                    this._easingToFrame(1.0, 1.0, locFrame);
                    locIsFindFrame = true;
                    break;
                } else if (locFrame.frameIndex * locUnitTime > time) {
                    if (j === 0) {
                        this._easingToFrame(1.0, 1.0, locFrame);
                        locIsFindFrame = false;
                    } else {
                        var locSrcFrame = locArray[j - 1];
                        var locDuration = (locFrame.frameIndex - locSrcFrame.frameIndex) * locUnitTime;
                        var locDelaytime = time - locSrcFrame.frameIndex * locUnitTime;
                        this._easingToFrame(locDuration, 1.0, locSrcFrame);
                        this._easingToFrame(locDuration, locDelaytime / locDuration, locFrame);
                        locIsFindFrame = true;
                    }
                    break;
                }
            }
        }
        return locIsFindFrame;
    },

    _easingToFrame: function (duration, delayTime, destFrame) {
        var action = destFrame.getAction(duration);
        var node = this.getActionNode();
        if (action == null || node == null)
            return;
        action.startWithTarget(node);
        action.update(delayTime);
    },

    /**
     * Returns if the action is done once time.
     * @returns {Boolean} that if the action is done once time
     */
    isActionDoneOnce: function () {
        if (this._action === null)
            return true;
        return this._action.isDone();
    }
});